src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs

```diff
@@ -2998,5 +2998,67 @@ public static bool IsFunctionPointerCallingConventionContext(this SyntaxTree syn
                    targetToken.Parent is FunctionPointerTypeSyntax functionPointerType &&
                    targetToken == functionPointerType.AsteriskToken;
         }
+
+        public static bool IsBaseClassContext(this SyntaxTree syntaxTree, SyntaxToken targetToken)
+        {
+            var possibleBaseListSyntax = targetToken.GetAncestor<BaseListSyntax>();
+            var possibleBaseTypeSyntax = targetToken.GetAncestor<BaseTypeSyntax>();
+
+            if (possibleBaseListSyntax is null ||
+                !possibleBaseListSyntax.IsParentKind(SyntaxKind.ClassDeclaration))
+            {
+                return false;
+            }
+
+            // We are in a non-written base type, e.g. "class C : $$"
+            if (possibleBaseTypeSyntax is null)
+            {
+                return targetToken.IsKind(SyntaxKind.ColonToken);
+            }
+
+            // 1. Current base type must be the first in the base list, because class inheritance is valid only at the first position in the base list
+            // 2. The only available target token type is qualified names. If there is generic names or whatever else,
+            // then it is not directly inheritance context and our restrictions are not valid there
+            return possibleBaseListSyntax.Types[0].Equals(possibleBaseTypeSyntax) &&
+                   targetToken.Parent is QualifiedNameSyntax { Parent: BaseTypeSyntax };
+        }
+
+        public static bool IsBaseEnumContext(this SyntaxTree syntaxTree, SyntaxToken targetToken)
+        {
+            // We don't care about qualified names, generics or whatever else, because in base list of enum can only contain one item of byte, short etc.
+            return targetToken.GetAncestor<BaseListSyntax>() is { Parent: EnumDeclarationSyntax };
+        }
+
+        public static bool IsBaseInterfaceContext(this SyntaxTree syntaxTree, SyntaxToken targetToken)
```

Comment:
so i would like somewhere for there to be a doc comment explaining what BaseInterfaceContext means.  I interpretted it as 'the inheritance list of an interface declaration'.  I see now that it's 'a location in *some* inheritance list where only interfaces are allowed'.  Doc'ing on the context object is fine :)